# Copyright (c) 2018. Mount Sinai School of Medicine
#
# Licensed under the Apache License, Version 2.0 (the "License").
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

"""
Alignment-related processing rules rewritten to use fq2bam pipeline with optional GPU support.
"""

from os.path import basename, join

# Read method and GPU info from config
METHOD = config.get("method", "cpu")
NUM_GPUS = config.get("gpus", 1)  # fq2bam supports multiple GPUs; default 1

def _get_read_group_header(wildcards):
    """
    Construct read group string for fq2bam.
    """
    prefix_basename = basename(wildcards.prefix)
    if prefix_basename.startswith("normal"):
        sample = "normal"
    elif prefix_basename.startswith("tumor"):
        sample = "tumor"
    else:
        raise ValueError("Unexpected prefix, cannot extract SM tag: %s" % wildcards.prefix)
    patient_id = config["input"]["id"]
    sample_id = "%s_%s" % (patient_id, sample)
    library = sample_id
    return "\\t".join([
        "@RG",
        "ID:%s" % wildcards.prefix,
        "SM:%s" % sample_id,
        "LB:%s" % library,
        "PL:Illumina"
    ])

################################################################
# fq2bam alignment
################################################################

rule fq2bam_single_end:
    input:
        r = join(WORKDIR, "{prefix}.fastq.gz"),
        ref = config["reference"]["genome"] + ".done"
    output:
        temp(join(WORKDIR, "{prefix}_aligned_coordinate_sorted_dups.bam"))
    params:
        reference = config["reference"]["genome"],
        extra = "-O 6 -E 1 -B 4"
    threads: _get_threads_for_alignment
    benchmark:
        join(BENCHMARKDIR, "{prefix}_fq2bam.txt")
    log:
        join(LOGDIR, "{prefix}_fq2bam.log")
    run:
        if METHOD == "gpu":
            shell(
                "fq2bam align "
                "--reference {params.reference} "
                "--input {input.r} "
                "--output {output} "
                "--threads {threads} "
                "--gpus {NUM_GPUS} "
                "> {log} 2>&1"
            )
        else:
            shell(
                "fq2bam align "
                "--reference {params.reference} "
                "--input {input.r} "
                "--output {output} "
                "--threads {threads} "
                "> {log} 2>&1"
            )

rule fq2bam_paired_end:
    input:
        r1 = join(WORKDIR, "{prefix}_R1.fastq.gz"),
        r2 = join(WORKDIR, "{prefix}_R2.fastq.gz"),
        ref = config["reference"]["genome"] + ".done"
    output:
        temp(join(WORKDIR, "{prefix}_aligned_coordinate_sorted_dups.bam"))
    params:
        reference = config["reference"]["genome"],
        extra = "-O 6 -E 1 -B 4"
    threads: _get_threads_for_alignment
    benchmark:
        join(BENCHMARKDIR, "{prefix}_fq2bam.txt")
    log:
        join(LOGDIR, "{prefix}_fq2bam.log")
    run:
        if METHOD == "gpu":
            shell(
                "fq2bam align "
                "--reference {params.reference} "
                "--input1 {input.r1} --input2 {input.r2} "
                "--output {output} "
                "--threads {threads} "
                "--gpus {NUM_GPUS} "
                "> {log} 2>&1"
            )
        else:
            shell(
                "fq2bam align "
                "--reference {params.reference} "
                "--input1 {input.r1} --input2 {input.r2} "
                "--output {output} "
                "--threads {threads} "
                "> {log} 2>&1"
            )

################################################################
# Merge aligned BAM fragments
################################################################

rule merge_normal_aligned_fragments:
    input:
        expand(join(WORKDIR, "normal_{fragment_id}_aligned_coordinate_sorted_dups.bam"),
               fragment_id=_get_fragment_ids("normal"))
    output:
        temp(join(WORKDIR, "normal_merged_aligned_coordinate_sorted_dups.bam"))
    threads: _get_half_cores
    run:
        if len(input) > 1:
            shell("sambamba merge -t {threads} {output} {input}")
        else:
            shell("cp {input} {output}")

rule merge_tumor_aligned_fragments:
    input:
        expand(join(WORKDIR, "tumor_{fragment_id}_aligned_coordinate_sorted_dups.bam"),
               fragment_id=_get_fragment_ids("tumor"))
    output:
        temp(join(WORKDIR, "tumor_merged_aligned_coordinate_sorted_dups.bam"))
    threads: _get_half_cores
    run:
        if len(input) > 1:
            shell("sambamba merge -t {threads} {output} {input}")
        else:
            shell("cp {input} {output}")
